### PR TITLE
Recover an agent whose container Docker can no longer serve logs for

### PR DESCRIPTION
**Pull request by devxoul**
## Background

An operator running `typeclaw compose restart` across three agents saw two fail:

```
▲  restarted 1/3 (2 failed)
  <agent-a>  ✖ failed: stop failed: Could not archive logs for container <id>: docker logs exited with code 1: can not get logs from container which is dead or marked for removal. Preserving the Docker container as the source of truth.
  <agent-b>  ✖ failed: stop failed: Could not archive logs for container <id>: docker logs exited with code 1: can not get logs from container which is dead or marked for removal. Preserving the Docker container as the source of truth.
  <agent-c>  ✔ restarted on host port 8973
```

`docker logs` returns that string verbatim from moby's `daemon/logs.go` whenever the daemon has marked the container `RemovalInProgress` or `Dead`. It rejects the request before ever consulting the logging driver, so the logs are permanently unretrievable — `--since` cannot narrow it — while `docker inspect` still reports `State.Running=false` as if nothing were wrong.

`archiveContainerLogs` (`src/container/log-archive.ts`) turned any non-zero `docker logs` exit into one generic failure, and `stop()` (`src/container/stop.ts`) treated any archive failure as fatal, preserving the container "as the source of truth." In this specific state nothing was being preserved: the daemon had already destroyed the only log source before `stop` even ran. The guard blocked the workflow while protecting nothing, and the message it printed blamed log archival for what was actually a wedged container record.

It was also unrecoverable through TypeClaw's own surface. `stop` died at the archive and therefore never issued `docker rm` at all; `restart` never reached `start`; and a standalone `start` ran the identical archive-before-remove hook through `cleanupRunCorpse` and failed the same way. The only escape was a manual host-side `docker rm -f`, which the error message never mentioned. This is the same shape AGENTS.md already names under "Security serves the product, not the other way around": `69401ed4 → 79bd56bf` fail-closed on ordinary Git debris, and `035da123 → 7b12232e` had a remediation that itself ran through the blocked surface.

## Summary

- **Classify, don't broaden** (`log-archive.ts`). `DockerLogArchiveResult`'s failure branch is now discriminated: `kind: 'unavailable'` for that one fixed daemon string, `kind: 'failed'` for everything else — invalid input, unresolved binary, timeout, overflow, termination bound, and every filesystem error. Archival still never reports success for a snapshot it did not write. The matched string is a fixed daemon protocol token, not human language, so it's exempt from the multi-language matching rule the same way `NO_REPLY` and GitHub `@login` are.
- **Tolerate only `unavailable`.** `stop` and `start`'s `archiveBeforeRemove` warn via `dockerLogsUnavailableWarning()` and continue to the ordinary non-force `docker rm`. A `failed` archive still preserves the container with the existing message, and `stop` still issues no `rm` at all in that case. **This is the actual fix**: reaching `docker rm` at all is what lets a `dead` container be reclaimed once its busy mount clears.
- **Diagnose instead of escalating** (`verify-running.ts`). Removal stays non-force and ID-keyed. New `describeUnremovableContainer` reads `.State.Status` through the existing `probeContainer`, and appends a state-specific remedy to the failure reason. It returns `null` for every other status, so those paths keep their reason byte-for-byte.
- **Never silent.** `stop` gains `onWarning` (mirroring `start`'s), threaded through `restart`, `compose stop`, `compose restart`, and the four CLI commands, which print collected warnings after the spinner settles, on both the success and failure paths.
- **Docs.** `AGENTS.md` and the two `docs/content/docs/reference/` pages stated flatly that any capture/retention failure preserves the container; all three now name the exemption and the non-force guarantee.
- Also lands one independent test-coverage commit for the pre-existing `cleanupRunCorpse` drain-timeout path, which had none.

## Why no `docker rm --force`

An earlier revision of this branch escalated to `docker rm --force` after ordinary removal failed, gated on a `Dead` re-probe. **It was removed, because moby shows it is a no-op in both states this bug can be in:**

- **`RemovalInProgress`** — `daemon/delete.go` runs `if inProgress := ctr.State.SetRemovalInProgress(); inProgress { return errdefs.Conflict(...) }` *before* any force handling, and `daemon/delete_test.go` asserts that conflict "even when forced" (`ContainerRmConfig{ForceRemove: true}`). Force cannot enter cleanup. The flag is also `json:"-"` and omitted from the struct built in `daemon/inspect.go`, so it is invisible to `{{.State.Dead}}` anyway.
- **`Dead`** — force only changes the branch for a **running** container (`daemon/delete.go`: `if ctr.State.IsRunning() { if !config.ForceRemove { conflict }; kill }`). A dead container is not running, so `docker rm` and `docker rm --force` take the identical `cleanupContainer` path — the same `containerfs.EnsureRemoveAll(ctr.Root)` retry that already failed.

Adding it would have weakened the deliberately non-force, ID-keyed removal invariant AGENTS.md marks load-bearing, in exchange for nothing. So the branch keeps removal exactly as it was and spends the effort on telling the operator what to actually do:

- `removing` → the removal is not draining; `docker rm`, including `--force`, cannot re-enter an in-flight removal; restart Docker/OrbStack, after which moby's restore path converts the container to `dead` and TypeClaw removes it on the next lifecycle command.
- `dead` → Docker could not remove the filesystem; resolve the busy mount or restart Docker, then retry. The underlying `docker rm` stderr is retained alongside it.

## What this does NOT do

- Adds no `--force` anywhere. `rg "'--force'" src/container/` matches only negative assertions in tests.
- Leaves `src/container/shared.ts` byte-identical to `main`.
- Does not change behavior for an ordinary `failed` archive — that path still preserves the container exactly as before.
- Does not change any reason string on a path where the status probe is unreadable.
- Does not touch `docker logs` retry/backoff, `--since` windowing, or retention pruning.

## Honest scope

For a container in `dead`, this restores recovery: `stop` now reaches `docker rm`, which retries cleanup. For a container in `removing`, nothing in-band can help — moby's deferred `ResetRemovalInProgress()` never runs while the removal is hung. That case turns from a misleading, unactionable archive error into an accurate one naming the daemon restart. Both beat the status quo, but only the first is a true auto-recovery.

## Review notes

- `isDockerLogsPermanentlyUnavailable` matches one fixed moby string case-insensitively — if that string changes upstream this falls back to `failed` (the safe direction) rather than mis-tolerating.
- `describeUnremovableContainer` lives in `verify-running.ts` because `shared.ts` cannot import it without a cycle (`verify-running.ts` already imports `DockerExec` from `shared.ts`). `cleanupRunCorpse` therefore keeps returning `'stuck'` and `start.ts` enriches the message at the preflight site, where a trustworthy container ID is in hand.

## Verification

- `bun run typecheck`, `bun run lint` (0 errors), `bun run format:check` all clean.
- Full suite: 13568 pass / 32 skip / 0 fail across 613 files.
- Each of the 8 commits verified independently — typecheck and tests green at every commit, not just at HEAD.
- Rebased onto current `main` (includes #1448, which also touches `start.ts`).
- Both directions covered: the threat still fails (a `failed` archive issues no `rm`; no path issues `-f`/`--force`; an unreadable probe preserves the original reason exactly), and the authorized workflow is restored (`unavailable` warns and proceeds to non-force removal, and the stale-corpse path then launches the replacement).

### CODE REVIEW

File: src/container/start.ts [hunk 1]
```diff
@@ -271,12 +271,19 @@ async function runStart({
...
-      await archiveLogs({
+    const archiveBeforeRemove = async (containerId: string) => {
+      const archive = await archiveLogs({
         agentDir: cwd,
         containerId,
         retentionDays: (await loadTypeclawConfig(cwd)).logs.retentionDays,
       })
+      if (archive.ok) return { ok: true as const }
+      if (archive.kind === 'failed') return { ok: false as const, reason: archive.reason }
+      const warning = dockerLogsUnavailableWarning(containerId)
```
[the remark this answers is not in the document]
Comment (devxoul) [reply src/container/start.ts:292]: Fixed in `01614851` (folded into the original commit). You were right, and the regression was mine: an earlier revision of this branch did pass `streamOutput: false` from the CLI, and I reverted that because `streamOutput` also gates `inheritStdio` for the Docker build (`start.ts` build path) — killing live build output would have been a worse regression. But I then wrongly concluded the stderr path was adequate, which left exactly the hole you describe: the CLI supplies a collector *and* keeps the streaming default, so the collector stayed empty and the warning landed under a live spinner.

The dispatch now keys on the callback, not on `streamOutput`:

```ts
if (onWarning !== undefined) onWarning(warning)
else if (streamOutput) process.stderr.write(`${warning}\n`)
```

with a comment at the site explaining why the obvious `if (streamOutput)` form is wrong here, so it doesn't get "simplified" back.

Coverage added at the domain layer, both directions:
- `routes the unavailable-logs warning to onWarning on the streaming CLI default` — constructs the exact production shape (`streamOutput` left at its `true` default + collector supplied), asserts the collector receives it and that nothing matching the warning reached `process.stderr`.
- `writes the unavailable-logs warning to stderr when no collector is supplied` — pins the fallback.

Mutation-checked: reverting the dispatch to the old `if (streamOutput) … else …` makes the first test fail (`216 pass / 1 fail`), so it genuinely guards the fix rather than restating it.

On "warning rendering after the spinner settles on success and failure" — that ordering is already in `src/cli/start.ts` and `src/cli/restart.ts` (`reportConfigWarnings(warnings)` runs after `s.stop(...)` on success and after `s.error(...)` before `process.exit(1)` on failure); this fix is what makes it actually reachable. I did not add a CLI-level test for it because there's no existing seam — `resolveController()` is called directly inside the citty `run` and there are no `src/cli/start.test.ts` / `stop.test.ts` today; testing it would mean stubbing clack and the controller module, which AGENTS.md's testing philosophy explicitly steers away from ("If you're mocking `@clack/prompts` … the domain logic is in the wrong place"). Happy to add one if you'd rather I introduce that seam.

One related thing I deliberately left alone: `githubCliWarning` a few lines below still dispatches on `streamOutput`. Its behavior is unchanged by this PR (before it, `src/cli/start.ts` passed no `onWarning` at all, so `streamOutput: true` → stderr was the only path), so fixing it here would be an unrelated behavior change. Say the word and I'll do it in this PR or a follow-up.

File: src/container/start.ts [hunk 2]
```diff
@@ -271,12 +276,19 @@ async function runStart({
...
+      if (streamOutput) process.stderr.write(`${warning}\n`)
```
[the remark this answers is not in the document]
Comment (devxoul) [reply src/container/start.ts]: Same root cause as the thread on line 287 — fixed in `01614851`, folded into the original commit.

`archiveBeforeRemove` now dispatches on the callback rather than on `streamOutput`, so a supplied `onWarning` wins even while streaming, and stderr is used only when no callback exists. Added `routes the unavailable-logs warning to onWarning on the streaming CLI default`, which builds the real production shape (streaming default + collector) and asserts the warning does **not** reach `process.stderr`, plus a companion test pinning the no-collector stderr fallback. Mutation-checked against the old dispatch.

Full reasoning, including why I didn't add a CLI-level test and why `githubCliWarning` was left alone, is in the reply on the line-287 thread.

File: src/container/start.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (devxoul) [reply src/container/start.ts:292]: Done — CLI-level coverage added in `261b792a` (folded into "Print lifecycle warnings once the spinner settles"). You were right that the container-level test only proved callback routing, not the CLI's rendering order.

There was no seam to test through: the logic lived inside the citty `run({ args })` closure calling `resolveController()` directly. So I introduced one the way `src/cli/status.ts` already does it — export the orchestration, keep the citty `run` a thin shell — and asserted on the emitted event sequence, per AGENTS.md §3 ("make orchestrators emit observable events ... and assert on the sequence").

`runStartCommand` / `runRestartCommand` take `{ start | restart, onEvent }` and never touch `spinner()`, `console.*`, or `process.exit`; the citty `run` maps events onto the real spinner and exit code. No behavior, message text, or ordering changed.

New `src/cli/start.test.ts` and `src/cli/restart.test.ts` assert the full sequences:

- start success — `spinner-start → spinner-stop → collected warnings → dockerfile warnings → success`
- start failure — `spinner-start → spinner-error → collected warnings`, and no `success` event
- restart success — `stop-start → stop-stop → start-start → start-stop → collected warnings → dockerfile warnings → success`
- restart failure before `onStopped` — errors the **stop** spinner
- restart failure after `onStopped` — errors the **start** spinner

Each fake also asserts `'streamOutput' in receivedOptions === false`, pinning the default streaming path you asked about — and guarding against re-introducing the `streamOutput: false` regression, which would silently kill live `docker build` output since that flag also gates `inheritStdio`.

Mutation-checked rather than assumed: moving the warnings emission before the spinner stop/error fails the corresponding test every time (e.g. start-success goes `1 pass / 1 fail`), then passes again on restore.

Also replaced a detached `start: resolveController().start` with a call-through closure over a single `resolveController()` — safe today since neither controller uses `this`, but a latent footgun if one ever does.

`bun run typecheck` / `lint` (0 errors) / `format:check` clean; full suite `13575 pass / 32 skip / 0 fail` across 615 files.

Note the earlier `flake guard` red was `startDaemon > GC ignores a threshold probe that raced a re-registration` in `src/hostd/daemon.test.ts` — a 30s timeout in a test this PR doesn't touch (no hostd files changed); it passed on re-run.

Leaving this thread open for you to close.
